### DISCUSSION TITLE
Interaction between Audio Resampling and touchRead

### DISCUSSION THREAD

**Discussion by HobmanX** [pschatzmann/arduino-audio-tools#1511] [Q&A] [answered]
I am playing a sketch (ESP32-PICO-D4) generating sound from Arrays and both waves are resampled, one is a fixed step and othe has various steps selected by toucgRead pins. The problem I have is the sketch starts up and plays an initial tone, then when I touch a touch pad to select another frequency by changing to another stepsize, I get a break in the tone (short ms) and the tone selected plays OK and touching any other pad works OK from then on, there are no breaks. The problem seem to only be when the step change id quite large (maybe 1.5 to 2.0 etc). If I touch a pad first that has only a small step change there are no breaks and all plays OK.  It only happens if first touch pad has a large step ans problem is only that the first change starts with a break even the initial tone is already playing. When I continue pressing the next pads all plays without breaks.  I have tried various pins and different setups in the sketch. At compiling it tells me that 44% of memory is used by the sketch. My question therefore .... Is there any interaction with interrupts or timers etc between touch and audio.....Can both work together.....?
OR is there any difference in calculations depending on whether the stepsize is large or small.....
The initial tone is setup in setup section etc whereas the rest of the selected tones are run in loop as the program runs...?

**Comment by pschatzmann** [comment, marked as answer]
I guess the first touch read is doing some initialization which takes more time then the subsequent calls.
Why don't you just issue a first call in the setup where you ignore the result, to "warm up" the processing ?

The other option would be to increase the i2s buffer, so that there is still audio data during this time gap.

**Reply by HobmanX** [reply]
I got it sorted....... I tried various options calling it at end of Setup and taking some thing out of loop to speed the time within loop but not the answer. I had 8 touch pads in an array polled via a for-next loop. The sketch started OK and played through setup and into loop and would be polling the touch pads waiting on some to be touched and the tones would be continuous no breaks. The for-next then detected the first not touched and then returned out of the for-next the number of the first pad not touched and played the corresponding tone, so if none touched then first tone is played, if the first 2 pads were touched then the 3rd tone (1st not touched) would be played and so on and all worked. The problem was if the first 3 or more were touched (as a first select) I would hear a click and see a short break and then the tone would play normally. Subsequent presses after that for any tone in that session all worked no matter what was pressed. So must have been some interaction between alot of touched pads at one time  and the audio on the first press, because any number of pads could be scanned after that and no problems. I scanned up the array and down the array and still same effect. None of that sorted it.
Eventually what I did was instead of scanning and returning on detection of first not pressed, I scanned the whole array and if it was detected as touched then set the bits in a byte for each not touched, then a for-next to scan the byte for the first 1 in it and play the corresponding tone....... That worked.... must be it meant the touchReads were over and saved in the byte,  and the playing the tone was seperated from the touchReads and step change happening when the byte bits were being read from the byte..
